### DISCUSSION TITLE
multy-objective in routing

### DISCUSSION THREAD

**Discussion by zengbin5966** [google/or-tools#2589] [Routing (and legacy CP) questions] [answered]
how can i customize my objective function without dimension?  Can I adjust the calculation method of cost in SolveMatchingModel?

**Comment by lperron** [comment, marked as answer]
Short answer. The objective is locked to be arc cost evalator + soft penalties + disjunction penalties + a few more.

**Reply by Roosjeva** [reply]
Can you change what is taken into account in the objective? Because i only want to calculate on used vehicles, distance and dropped nodes. And i do want to have time as constraint but does not have to be calculated in the objective

**Reply by lperron** [reply]
re-read my answer :-)

**Reply by Roosjeva** [reply]
So no it is not possible?  
only by letting out the constraints

**Reply by Roosjeva** [reply]
Then other question, it seems that the script is better working then the optimal solution obtained in literature. How can this happen, can i assume that or-tools works better then the used algorithm in literature?

**Reply by lperron** [reply]
Just do not set the arc cost evaluator to the time dimension.

**Reply by Roosjeva** [reply]
Okay so if I have the lines below the objective would be:
distances + penalties + used vehicles ? 

routing.AddDisjunction([manager.NodeToIndex(node)], penalty)
routing.SetFixedCostOfVehicle(pen, i)
routing.SetArcCostEvaluatorOfAllVehicles(distance_callback_index)

Are there other functions/variables used in or-tools to add to the objective ?

**Reply by Mizux** [reply]
You can't have a custom cost function, BUT the cost function can integrate:

- [mandatory] The sum of all arc cost used (used by any initial first strategy)
  ref: https://github.com/google/or-tools/blob/b37d9c786b69128f3505f15beca09e89bf078a89/ortools/constraint_solver/routing.h#L901-L906
- [optional] The span cost of any dimension (e.d. one coefficient per vehicle, default 0)
- [optional] The global span cost of any dimension (e.d. one coefficient per dimension, default 0)
- [optional] A vehicle fixed cost (e.d. one cost per vehicle, default 0)
  ref: https://github.com/google/or-tools/blob/b37d9c786b69128f3505f15beca09e89bf078a89/ortools/constraint_solver/routing.h#L907-L911
- [optional] A penalty cost for any dropping node using AddDisjunction() (default all nodes are mandatory)
  ref: https://github.com/google/or-tools/blob/b37d9c786b69128f3505f15beca09e89bf078a89/ortools/constraint_solver/routing.h#L606-L624 
- [optional] A linear penalty cost per node if using SoftCumulVar Lower/Upper Cost (to simulate earliness/tardiness) (default only hard limit [0, vehicle max capacity aka horizon of the dimension])
  ref: https://github.com/google/or-tools/blob/b37d9c786b69128f3505f15beca09e89bf078a89/ortools/constraint_solver/routing.h#L2514-L2523
  ref: https://github.com/google/or-tools/blob/b37d9c786b69128f3505f15beca09e89bf078a89/ortools/constraint_solver/routing.h#L2536-L2546

Misc/Not tested yet:
- [optional] Soft constraint a group of node to be visited by the same vehicle
  ref: https://github.com/google/or-tools/blob/b37d9c786b69128f3505f15beca09e89bf078a89/ortools/constraint_solver/routing.h#L675-L679

note: You can search for "soft" which means add extra cost to objective if not fulfilled while hard constraint must be respected... 

ref: https://groups.google.com/g/or-tools-discuss/c/YCUNTmWr8Us

**Reply by Roosjeva** [reply]
Thank you Mizux your detailed explanation helps a lot. 
To make your explanation practice, I have the following question:
- If I want to optimize on distance and time, should I turn on a set arc cost evaluator for both? 

So as follows:

routing.AddDisjunction([manager.NodeToIndex(node)], penalty)
 routing.SetArcCostEvaluatorOfAllVehicles(distance_callback_index)
 routing.SetArcCostEvaluatorOfAllVehicles(time_callback_index)

pen = 1000
    for i in range(data['num_vehicles']):  
        routing.SetFixedCostOfVehicle(pen, i)

**Reply by Mizux** [reply]
IF I recall correctly the second call will "replace" the first one aka the distance_callback is dropped and won't participate to the objective cost...
**BUT** you can use the Dimension span cost, i.e. by default each dimension won't participate to the objective aka arc value are just use to model some constraints like (e.g. max working time)
if you want to "integrate" dimension CumulVar() aka arc value to the objective, simply set the dimension span cost to 1 (ed you can see it to zero by default)
ref: https://github.com/google/or-tools/blob/2cb85b4eead4c38e1c54b48044f92087cf165bce/ortools/constraint_solver/routing.h#L2680-L2687
